README.md

```diff
@@ -18,36 +18,203 @@ testing first appeard in Erlang's proprietary QuickCheck. The
+we can only read from references that already have been created. The
+pre-conditions are used while generating programs (lists of actions).
+
+```haskell
+precondition :: Model Symbolic -> Action Symbolic resp -> Bool
+precondition _         New           = True
+precondition (Model m) (Read  ref)   = ref `elem` map fst m
+precondition (Model m) (Write ref _) = ref `elem` map fst m
+precondition (Model m) (Inc   ref)   = ref `elem` map fst m
+```
+
+The transition function explains how actions change the model. Note that the
+transition function is polymorphic in `v`. The reason for this is that we use
+the transition function both while generating and executing.
+
+```haskell
+transition :: Model v -> Action v resp -> v resp -> Model v
+transition (Model m) New           ref = Model (m ++ [(Reference ref, 0)])
+transition m         (Read  _)     _   = m
+transition (Model m) (Write ref i) _   = Model (update ref i         m)
+transition (Model m) (Inc   ref)   _   = Model (update ref (old + 1) m)
+  where
+  Just old       = lookup ref m
+
+update :: Eq a => a -> b -> [(a, b)] -> [(a, b)]
+update ref i m = (ref, i) : filter ((/= ref) . fst) m
+```
+
+Post-conditions are checked after we executed an action and got access to the
+result.
+
+```haskell
+postcondition :: Model Concrete -> Action Concrete resp -> resp -> Property
+postcondition _         New         _    = property True
+postcondition (Model m) (Read ref)  resp = lookup ref m === Just resp
+postcondition _         (Write _ _) _    = property True
+postcondition _         (Inc _)     _    = property True
+```
+
+Finally, we have to explain how to generate and shrink actions.
+
+```haskell
+generator :: Model Symbolic -> Gen (Untyped Action)
+generator (Model m)
+  | null m    = pure (Untyped New)
+  | otherwise = frequency
+      [ (1, pure (Untyped New))
+      , (8, Untyped .    Read  <$> elements (map fst m))
+      , (8, Untyped <$> (Write <$> elements (map fst m) <*> arbitrary))
+      , (8, Untyped .    Inc   <$> elements (map fst m))
+      ]
+
+shrinker :: Action v resp -> [Action v resp]
+shrinker (Write ref i) = [ Write ref i' | i' <- shrink i ]
+shrinker _             = []
+```
 
-┌──────────────────────┐
-│ New                  │
-│                 ⟶ $0 │
-└──────────────────────┘
-            │ ┌────────┐
-            │ │ Inc $0 │
-┌─────────┐ │ │        │
-│ Inc $0  │ │ │        │
-│         │ │ │   ⟶ () │
-│         │ │ └────────┘
-│    ⟶ () │ │
-└─────────┘ │
-┌─────────┐ │
-│ Read $0 │ │
-│     ⟶ 1 │ │
-└─────────┘ │
+We can now define a sequential and a parallel property as follows.
+
+```haskell
+prop_references :: Problem -> Property
+prop_references prb = forAllProgram
+  generator
+  shrinker
+  precondition
+  transition
+  initModel $ \prog ->
+    runAndCheckProgram
+      precondition
+      transition
+      postcondition
+      initModel
+      (semantics prb)
+      ioProperty
+      prog
+
```

Comment:
Maybe move the parallel property definition below the running of the sequential, motivating race condition...